### PR TITLE
BSIP87 Tests

### PR DESCRIPTION
**Pull request by christophersanborn**
Unit tests followup to #2151.

### CODE REVIEW

**Comment by abitmore** [comment]
There are commits
* Improve error message in asset_claim_fees_evaluator
* Revert "Improve error message in asset_claim_fees_evaluator"

Please rebase and remove these commits from the history.

Thanks.

**Comment by abitmore** [comment]
OK, I changed my mind. The 2 commits are quite small, but it might need quite some efforts to remove them, I think it's fine to keep them, although generally it's not perfect.

File: tests/tests/bitasset_tests.cpp [hunk 1]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+
+         // Obtain asset object after a block is generated to obtain the final object that is commited to the database
+         generate_block(); trx.clear(); set_expiration(db, trx);
+         const asset_object &smartbit = get_asset("SMARTBIT");
+         const asset_bitasset_data_object& smartbit_bitasset_data = (*smartbit.bitasset_data_id)(db);
+         // Confirm that the asset is to be backed by JCOIN
+         BOOST_CHECK(smartbit_bitasset_data.options.short_backing_asset == jillcoin.id);
+
+         // Fund balances of the actors
+         issue_uia(alice, jillcoin.amount(5000 * std::pow(10, jillcoin.precision)));
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp:1454]: Please avoid using `std::pow()` since it uses floating points thus is not deterministic. Same below (multiple).

File: tests/tests/bitasset_tests.cpp [hunk 2]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+
+         /**
+          * Claim all of the available collateral asset fees
+          */
+         trx.clear();
+         claim_op.amount_to_claim = accumulation_amount;
+         trx.operations.push_back(claim_op);
+         set_expiration(db, trx);
+         sign(trx, smartissuer_private_key);
+         PUSH_TX(db, trx);
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp:1526]: Better check balance and dynamic asset data after the transaction.

Also need test cases for
* claiming less than available, and
* claiming more than available.

File: tests/tests/bitasset_tests.cpp [hunk 3]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+          * Cause some collateral of JCOIN to be accumulated as collateral fee within the SMARTBIT asset type
+          */
+         // HACK: Before BSIP74 or BSIP87 are introduced, it is not formally possible to accumulate collateral fees.
+         // Therefore, the accumulation for this test will be informally induced by direct manipulation of the database.
+         // More formal tests will be provided with the PR for either BSIP74 or BSIP87.
+         // IMPORTANT: The use of this hack requires that no additional blocks are subsequently generated!
+         asset accumulation_amount = jillcoin.amount(40 * std::pow(10, jillcoin.precision)); // JCOIN
+         db.adjust_balance(alice.id, -accumulation_amount); // Deduct 40 JCOIN from alice as a "collateral fee"
+         smartbit.accumulate_fee(db, accumulation_amount); // Add 40 JCOIN from alice as a "collateral fee"
+
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp:1491]: Better check balances and dynamic asset data after the modification.

File: tests/tests/bitasset_tests.cpp [hunk 4]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+          */
+         trx.clear();
+         asset_claim_fees_operation claim_op;
+         claim_op.issuer = smartissuer.id;
+         claim_op.extensions.value.claim_from_asset_id = smartbit.id;
+         claim_op.amount_to_claim = jillcoin.amount(5 * std::pow(10, jillcoin.precision));
+         trx.operations.push_back(claim_op);
+         set_expiration(db, trx);
+         sign(trx, smartissuer_private_key);
+         GRAPHENE_REQUIRE_THROW(PUSH_TX(db, trx), fc::exception);
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp:1471]: Need a test case to verify that it's unable to wrap an `asset_claim_fees_operation` with a non-null `extensions.value.claim_from_asset_id` in a `proposal_create_operation`.

File: tests/tests/bitasset_tests.cpp [hunk 5]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+         change_backing_asset_op.asset_to_update = smartbit.id;
+         change_backing_asset_op.issuer = smartissuer.id;
+         change_backing_asset_op.new_options.short_backing_asset = izzycoin.id;
+         trx.operations.push_back(change_backing_asset_op);
+         sign(trx, smartissuer_private_key);
+         REQUIRE_EXCEPTION_WITH_TEXT(PUSH_TX(db, trx), "Must claim collateral-denominated fees");
+
+
+         /**
+          * Attempt to claim a negative amount of the collateral asset fees.
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp:1507]: Attempt to claim `0` too? Perhaps there is already a test case for it elsewhere, I didn't check.

File: tests/tests/force_settle_fee_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,1105 @@
...
+    * @param issuer Issuer ID
+    * @param force_settlement_offset_percent Force-settlement offset percent
+    * @param force_settlement_fee_percent Force-settlement fee percent (BSIP87)
+    * @return Asset object
+    */
+   const asset_object &create_smart_asset(
+           const string &name,
+           account_id_type issuer /* = GRAPHENE_WITNESS_ACCOUNT */,
+           uint16_t force_settlement_offset_percent /* 100 = 1% */,
+           optional<uint16_t> force_settlement_fee_percent /* 100 = 1% */
```
Comment (abitmore) [inline tests/tests/force_settle_fee_tests.cpp:73]: Ideally, we define a default value for this argument, then the function above with one argument fewer can be removed. Current code is acceptable for test cases though.

BTW I think it's better to add new parameters to the functions in `database_fixture` class, rather than adding new functions here. It's not a priority though, we can refactor in the future.

File: tests/tests/force_settle_fee_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,1105 @@
...
+         if (issuer == GRAPHENE_WITNESS_ACCOUNT)
+            flags |= witness_fed_asset;
+         creator.common_options.issuer_permissions = flags;
+         creator.common_options.flags = flags & ~global_settle;
+         creator.common_options.core_exchange_rate = price(asset(1, asset_id_type(1)), asset(1));
+
+         creator.bitasset_opts = bitasset_options();
+         creator.bitasset_opts->force_settlement_offset_percent = force_settlement_offset_percent;
+         creator.bitasset_opts->short_backing_asset = backing_asset;
+         if (force_settlement_fee_percent.valid()) {
```
Comment (abitmore) [inline tests/tests/force_settle_fee_tests.cpp:99]: This check is unnecessary.

File: tests/tests/force_settle_fee_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,1105 @@
...
+   BOOST_AUTO_TEST_CASE(force_settle_fee_1_test) {
+      try {
+         ///////
+         // Initialize the scenario
+         ///////
+         // Get around Graphene issue #615 feed expiration bug
+         generate_blocks(HARDFORK_615_TIME);
+         generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+         // Advance to the when the force-settlement fee activates
```
Comment (abitmore) [inline tests/tests/force_settle_fee_tests.cpp:145]: Ideally we need test cases to test that operations with new extensions (as well as being nested in proposals) aren't allowed before the hard fork.

File: tests/tests/force_settle_fee_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,1105 @@
...
+         transfer(committee_account, assetowner.id, asset(initial_balance_core));
+         transfer(committee_account, feedproducer.id, asset(initial_balance_core));
+         transfer(committee_account, michael_id, asset(initial_balance_core));
+         transfer(committee_account, paul.id, asset(initial_balance_core));
+
+         ///////
+         // 1. Create assets
+         ///////
+         const uint16_t usd_fso_percent = 5 * GRAPHENE_1_PERCENT; // 5% Force-settlement offset fee %
+         const uint16_t usd_fsf_percent_0 = 0 * GRAPHENE_1_PERCENT; // 0% Force-settlement offset fee %
```
Comment (abitmore) [inline tests/tests/force_settle_fee_tests.cpp:408]: Typo in comment. It's not for `offset`.

File: tests/tests/bitasset_tests.cpp [hunk 10]
```diff
@@ -1407,4 +1407,138 @@ BOOST_AUTO_TEST_CASE(hf_890_test_hf1270)
...
+          * Advance to when the collateral fee container is activated
+          */
+         generate_blocks(HARDFORK_CORE_BSIP_87_74_COLLATFEE_TIME);
+         generate_block(); trx.clear(); set_expiration(db, trx);
+
+
+         /**
+          * Cause some collateral of JCOIN to be accumulated as collateral fee within the SMARTBIT asset type
+          */
+         // HACK: Before BSIP74 or BSIP87 are introduced, it is not formally possible to accumulate collateral fees.
```
[the remark this answers is not in the document]
Comment (abitmore) [reply tests/tests/bitasset_tests.cpp:1484]: Update: another test case in `force_settle_fee_tests.cpp` addressed the concerns by checking the message in the exception.

File: tests/tests/force_settle_fee_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,1105 @@
...
+
+         // Check Paul's debt to the blockchain
+         // Yanna redeemed 10 bitUSD from the blockchain, and the blockchain closed this amount from Paul's debt to it
+         BOOST_CHECK_EQUAL(paul_initial_usd - rachel_settle_amount - michael_settle_amount - yanna_settle_amount,
+                           call_paul_id(db).debt.value);
+         // The call order has the original amount of collateral less what was redeemed by Rachel, Michael, Yanna
+         BOOST_CHECK_EQUAL(paul_initial_core - rachel_fso_remainder_core - michael_fso_remainder_core - yanna_fso_remainder_core,
+                           call_paul_id(db).collateral.value);
+
+         // The asset's force settlement fee % should still not be set
```
Comment (abitmore) [inline tests/tests/force_settle_fee_tests.cpp:686]: The comment is incorrect.

File: tests/tests/force_settle_fee_tests.cpp (hunk 6 above)
Comment (MichelSantos) [reply tests/tests/force_settle_fee_tests.cpp:73]: Those are reasonable points.  We'll defer the refactoring for now.

**Comment by MichelSantos** [comment]
Thanks for the helpful review @abitmore.  Hopefully the revisions address all of your comments.

File: tests/tests/force_settle_fee_tests.cpp (hunk 8 above)
Comment (abitmore) [reply tests/tests/force_settle_fee_tests.cpp:143]: Thank you @MichelSantos.
Only this hasn't been addressed. Commenting here, I meant the hardfork checks for `asset_create_operation` and `asset_update_bitasset_operation`.

Comment (abitmore) [reply tests/tests/force_settle_fee_tests.cpp:143]: Update: there is one test case for asset creation around line 411. I didn't find others.

**Comment by christophersanborn** [comment]
@abitmore — I still have one open question [here](https://github.com/bitshares/bitshares-core/pull/2151#issuecomment-626216509) as to whether there is merit in rejecting an operation from `do_evaluate()` if `force_settle_fee_parameter` is unset in the operation but is set to a non-zero value in the asset object.
